File: DP-4-Stealing/stealing.cpp
#include <bits/stdc++.h>

using namespace std;

int solution(vector<int> money)
{
    int length = money.size();
    int d1[length];
    int d2[length];
    memset(d1, 0, sizeof(d1));
    memset(d2, 0, sizeof(d2));
    d1[0] = d1[1] = money[0];
    d2[1] = money[1];
    for (int i = 2; i < length; i++)
    {
        d1[i] = max(d1[i - 1], d1[i - 2] + money[i]);
        d2[i] = max(d2[i - 1], d2[i - 2] + money[i]);
    }
    return max(d1[length - 2], d2[length - 1]);
}

int main(void)
{
    // cout << solution({1, 2, 3, 1}) << endl;
    cout << solution({10, 2, 2, 100, 2}) << endl;
    return 0;
}

File: DP-4-Stealing/stealing.py
def solution(m):
    d1 = [0] * len(m)
    d1[0] = m[0]
    d1[1] = 0
    d1[2] = max(d1[1], d1[0] + m[2])
    for i in range(3, len(m)):
        d1[i] = max(d1[i-1], d1[i-2] + m[i], d1[i-3] + m[i])
    print(d1)

    d2 = [0] * len(m)
    d2[0] = 0
    d2[1] = m[1]
    d2[2] = max(d2[1], d2[0] + m[2])
    for i in range(3, len(m)):
        d2[i] = max(d2[i-1], d2[i-2] + m[i], d2[i-3] + m[i])
    print(d2)

    return max(d1[-2], d2[-1])


# return 4
# print(solution([1, 2, 3000000, 1]))
print(solution([1, 2, 3, 1]))
# print(solution([10, 2, 2, 100, 2]))
# print(solution([1, 2, 3]))
